### ISSUE TITLE
Publish to crates.io

### ISSUE THREAD

**Issue by stefnotch** [GaloisInc/mir-json#48]
It'd be lovely if this tool were published to crates.io, like the majority of Rust tools.

Then the installation instructions might be simpler for users who already have Rust.

**Comment by RyanGlScott** [comment]
I agree. One reason that we haven't yet uploaded `mir-json` to `crates.io` is that any given `mir-json` release would only be buildable with a particular `rustc` nightly, and I could imagine lots of bug reports from confused users who try to install `mir-json` with an unsupported `rustc` version. That being said, we could include a [`rust-toolchain.toml`](https://rust-lang.github.io/rustup/overrides.html#the-toolchain-file) file to ensure that when a user runs `cargo install mir-json`, `cargo` will download the correct nightly before attempting to build the tool.

**Comment by RyanGlScott** [comment]
Actually, I'm not sure if `rust-toolchain.toml` takes effect with `cargo install` or not. It _would_ take effect if you cloned the `mir-json` repo and built it from source, which would simplify the current story slightly (in that you wouldn't need to manually change your Rust toolchain before building).

**Comment by stefnotch** [comment]
Ah, I see. I suppose a stable mir format would be needed for it to work with multiple rustc versions.

I personally would love anything that simplifies the installation instructions. Even getting rid of a single step is frequently worth it, in my opinion.

**Comment by RyanGlScott** [comment]
Ah, it turns out that we already have a `rust-toolchain.toml` file:

https://github.com/GaloisInc/mir-json/blob/3b538589852282a752e517b69fc61b627cd6d4bd/rust-toolchain.toml#L1-L2

I think that we need to add `components = ["rustc-dev"]` as well, but most of the ingredients are there.

Another thought that popped into my mind: there are several commands inside of `mir-json` that spawn `rustc` subprocesses. If a user installs `mir-json` with one `rustc` version (the intended version) but later changes their toolchain to have a different `rustc` version as the default, will `mir-json`'s `rustc` subprocesses still pick the intended version? Surprisingly, some experimentation suggests that the answer is "yes". I'm not sure how that works exactly, but that is a fortunate coincidence.

**Comment by spernsteiner** [comment]
> If a user installs `mir-json` with one `rustc` version (the intended version) but later changes their toolchain to have a different `rustc` version as the default, will `mir-json`'s `rustc` subprocesses still pick the intended version?  Surprisingly, some experimentation suggests that the answer is "yes".

I suspect this only works in certain cases.  The various `mir-json` binaries link against a specific version of the rustc libraries and will always behave like the nightly they're built against.  But we might run into trouble when invoking normal `rustc` instead of `mir-json` or one of our other binaries.  In particular, using `cargo crux-test` in a project that uses proc macros might compile the proc macros with the default toolchain and produce proc macro crates that can't be loaded by mir-json.

We could probably fix this by setting `RUSTUP_TOOLCHAIN` or passing `+nightly-xxx` args in more places.

**Comment by sand7000** [comment]
I want to make sure I understand the intent of this ticket. Here is my proposed solution, let me know if it checks all the boxes.

I will write an installation script that we can publish. The script will do the following:

1. Ensure the correct nightly (nightly-2023-01-23) is installed.
2. Install `mir-json` from crates.io using the correct nightly:

```
cargo +nightly-2023-01-23 install mir-json --root ~/.mir-json/
```

3. Create wrapper scripts:

 - For each installed binary (e.g., mir-json, cargo-saw-build), a corresponding wrapper script is written to ~/.cargo/bin/
 - Each wrapper sets RUSTUP_TOOLCHAIN=nightly-2023-01-23 and then delegates to the real binary in ~/.mir-json/bin

4. Detect and warn if a conflicting binary already exists in ~/.cargo/bin (e.g., from a previous cargo install) and install over it if desired.

This approach ensures that `cargo saw-build` works as expected, since Cargo discovers cargo-saw-build via $PATH and ~/.cargo/bin/ should already be on the PATH. The only issue with this is if people install using:

```
cargo install mir-json
```

It will not give an indication that the binaries aren't pinned to the correct nightly.

**Comment by spernsteiner** [comment]
It might be worth looking into how Kani handles this - as far as I know, Kani is still tied to a specific rustc nightly just like crux-mir is.  I looked at one point and IIRC `cargo install kani` would give you a little helper binary that downloads the correct toolchain and either downloads or builds the real Kani binary the first time it's run.

**Comment by RyanGlScott** [comment]
Indeed, Kani does have a crates.io upload (see the installation instructions [here](https://model-checking.github.io/kani/install-guide.html)) so that you can run `cargo install --locked kani-verifier`, but also requires you to run `cargo kani setup` to do the rest of the Kani-specific setup. That might be one viable way to get `mir-json` onto crates.io.

That being said, I don't know if we are actually in that much of a hurry to get onto crates.io just yet. There are some intermediate steps that we could do in the meantime to improve the user experience, e.g., the `RUSTUP_TOOLCHAIN` idea discussed in https://github.com/GaloisInc/mir-json/issues/48#issuecomment-1662534782. That should be doable without a substantial amount of effort to overhaul the installation story.

**Comment by RyanGlScott** [comment]
Rather than defining a separate script to define `RUSTUP_TOOLCHAIN`, we could just as well define the `RUSTUP_TOOLCHAIN` environment variable from the Rust code itself. For instance, various `mir-json` tools define the `RUSTC_WRAPPER` environment variable (see [here](https://github.com/GaloisInc/mir-json/blob/f336d18ccfc7853df2862fa89ea7d3343bf3a50a/src/bin/cargo_test_common.rs#L174) for an example) using `.env()`, so it seems viable to apply similar treatment to `RUSTUP_TOOLCHAIN`.

**Comment by qsctr** [comment]
Also, all the `mir-json` binaries are actually already wrapped by [`wrapper.rs`](https://github.com/GaloisInc/mir-json/blob/f336d18ccfc7853df2862fa89ea7d3343bf3a50a/src/bin/wrapper.rs), which sets up environment variables then runs the real binary. So that could be a place to set up `RUSTUP_TOOLCHAIN`.
